system/Publisher/Publisher.php

```diff
@@ -0,0 +1,736 @@
+	{
+		$this->files = [];
+
+		return $this->addFiles($files);
+	}
+
+	/**
+	 * Verifies and adds files to the list.
+	 *
+	 * @param string[] $files
+	 *
+	 * @return $this
+	 */
+	final public function addFiles(array $files)
+	{
+		foreach ($files as $file)
+		{
+			$this->addFile($file);
+		}
+
+		return $this;
+	}
+
+	/**
+	 * Verifies and adds a single file to the file list.
+	 *
+	 * @param string $file
+	 *
+	 * @return $this
+	 */
+	final public function addFile(string $file)
+	{
+		$this->files[] = self::resolveFile($file);
+
+		return $this;
+	}
+
+	/**
+	 * Removes files from the list.
+	 *
+	 * @param string[] $files
+	 *
+	 * @return $this
+	 */
+	final public function removeFiles(array $files)
+	{
+		$this->files = array_diff($this->files, $files);
+
+		return $this;
+	}
+
+	/**
+	 * Removes a single file from the list.
+	 *
+	 * @param string $file
+	 *
+	 * @return $this
+	 */
+	final public function removeFile(string $file)
+	{
+		return $this->removeFiles([$file]);
+	}
+
+	//--------------------------------------------------------------------
+
+	/**
+	 * Verifies and adds files from each
+	 * directory to the list.
+	 *
+	 * @param string[] $directories
+	 * @param bool $recursive
+	 *
+	 * @return $this
+	 */
+	final public function addDirectories(array $directories, bool $recursive = false)
+	{
+		foreach ($directories as $directory)
+		{
+			$this->addDirectory($directory, $recursive);
+		}
+
+		return $this;
+	}
+
+	/**
+	 * Verifies and adds all files from a directory.
+	 *
+	 * @param string  $directory
+	 * @param boolean $recursive
+	 *
+	 * @return $this
+	 */
+	final public function addDirectory(string $directory, bool $recursive = false)
+	{
+		$directory = self::resolveDirectory($directory);
+
+		// Map the directory to depth 2 to so directories become arrays
+		foreach (directory_map($directory, 2, true) as $key => $path)
+		{
+			if (is_string($path))
+			{
+				$this->addFile($directory . $path);
+			}
+			elseif ($recursive && is_array($path))
+			{
+				$this->addDirectory($directory . $key, true);
+			}
+		}
+
+		return $this;
+	}
+
+	//--------------------------------------------------------------------
+
+	/**
+	 * Verifies and adds paths to the list.
+	 *
+	 * @param string[] $paths
+	 * @param bool $recursive
+	 *
+	 * @return $this
+	 */
+	final public function addPaths(array $paths, bool $recursive = true)
+	{
+		foreach ($paths as $path)
+		{
+			$this->addPath($path, $recursive);
+		}
+
+		return $this;
+	}
+
+	/**
+	 * Adds a single path to the file list.
+	 *
+	 * @param string  $path
+	 * @param boolean $recursive
+	 *
+	 * @return $this
+	 */
+	final public function addPath(string $path, bool $recursive = true)
+	{
+		$full = $this->source . $path;
+
+		// Test for a directory
+		try
+		{
+			$directory = self::resolveDirectory($full);
+		}
+		catch (PublisherException $e)
+		{
+			return $this->addFile($full);
+		}
+
+		return $this->addDirectory($full, $recursive);
```

Comment:
This can be placed inside the `try`

Reply:
We want this to throw if `addDirectory()` or `addFile()` fail. The above `try..catch` is to determine whether the `$path` is really really a directory or not.